### ISSUE TITLE
Unable to install NiceGUI on 32 bits Windows since orjson dependency was added

### ISSUE THREAD

**Issue by ghuserlb** [zauberzeug/nicegui#518]
Hi,

I would like to mention that since v1.1.10 of nicegui and updates on plotly, a dependency has been introduced on orjson package. However this package is not available on 32 bits architectures (or at least for people running Python 32 bits) which is a serious limitation.

Do you have any plan on switching to another less constrained json library? Or maybe a more modular deployment based on components (nicegui-plotly, nicegui-quasar, ...)?

Thanks!

**Comment by falkoschindler** [comment]
Thanks for pointing this out, @ghuserlb!
This is the ticket which in introduced orjson: https://github.com/zauberzeug/nicegui/issues/380.
@rbeeli, do you have an idea how to deal with orjson on 32 bit machines? Can we automatically include/exclude it depending on the architecture?

**Comment by rodja** [comment]
Oh that is really serious. RaspberryPi is using 32 bit. We should support it. But it seems to be of no priority to orjson: https://github.com/ijl/orjson/issues/130

**Comment by rbeeli** [comment]
I would only consider it if the CI/CD pipeline would get extended to run tests on 32 bit systems, otherwise it's not maintainable, especially not if you add conditional module loading of a central component like the `json` module.
Other than that, in theory, it should be possible to have a different serializer for 32 bit systems, i.e. the Python default with some extensions for `numpy` support.

**Comment by rodja** [comment]
I think RaspberryPi support is very crucial. I think we can execute basic startup tests with https://github.com/wbond/pi-github-runner. The physical Pi`s could be hosted at Zauberzeug. How would we pick the right serializer for 32 bit systems?

**Comment by rbeeli** [comment]
Would be great setting up some automated test system for 32 bit architectures, if you want to officially support it, I think it's necessary, it would have caught this issue before the new release.

Two options to solve the issue:
1. Make `orjson` an extra, see [here](https://stackoverflow.com/a/13681679), so by default it would use Python's `json` module, unless we specify `nicegui[orjson]` when `pip` installing the package.
2. Always install `orjson`, consider it the default serializer, but fallback to Python's native `json` module if on 32 bit architecture. I don't think there can be conditional dependencies by architecture, i.e. only install `orjson` on 64 bit architectures?

Personally, I prefer option 2., because 99% of users are better off using `orjson`.

To not break the new Plotly component, we would need to teach the Python `json` module to correctly serialize `numpy` arrays, which should be possible as well, see [here](https://stackoverflow.com/a/49677241).

Opinions?

**Comment by rodja** [comment]
I also prefer option 2. We should have the best available performance by default. Could you create a pull request. We start working on a 32 bit platform test.

**Comment by rodja** [comment]
I just tested it myself on a Raspberry Pi. Latest NiceGUI installs fine but is showing the following error on start:

```
# python3 main.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ImportError: libcblas.so.3: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/test/main.py", line 1, in <module>
    from nicegui import ui
  File "/usr/local/lib/python3.9/dist-packages/nicegui/__init__.py", line 1, in <module>
    from . import elements, globals, ui
  File "/usr/local/lib/python3.9/dist-packages/nicegui/ui.py", line 44, in <module>
    from .elements.scene import Scene as scene
  File "/usr/local/lib/python3.9/dist-packages/nicegui/elements/scene.py", line 9, in <module>
    from .scene_object3d import Object3D
  File "/usr/local/lib/python3.9/dist-packages/nicegui/elements/scene_object3d.py", line 4, in <module>
    import numpy as np
  File "/usr/local/lib/python3.9/dist-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.9 from "/usr/bin/python3"
  * The NumPy version is: "1.24.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: libcblas.so.3: cannot open shared object file: No such file or directory
```

Which was not the error I've expected. I used the command

    sudo apt-get install libatlas-base-dev

to fix it as suggested in https://stackoverflow.com/a/56972113/364388. After that the latest NiceGUI 1.1.11 starts without errors. Please help us reproduce the issue. The system looks like this:

```bash
~/test# cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

~/test# uname -mrs
Linux 5.15.84-v7+ armv7l

~/test# getconf LONG_BIT
32
```

What is different?

**Comment by ghuserlb** [comment]
Hi all,

Thanks for the efforts! Afaik I'm working on a Windows python 3.7 32 bits setup, and for various reasons I cannot use a 64 bits python.

It looks like your raspberry pi runs a debian based distro and I suspect debian to provide specifically compiled python modules. Did you install orjson via pip install or apt get? I expect the pip install would fail,whereas the apt get might succeed...

It would be too bad to have to give up your great app on my work setup :(

Regards


-------- Message d'origine --------
De : Rodja Trappe ***@***.***>
Date : 15/03/2023 17:05 (GMT+01:00)
À : zauberzeug/nicegui ***@***.***>
Cc : ghuserlb ***@***.***>, Mention ***@***.***>
Objet : Re: [zauberzeug/nicegui] unable to install nicegui on 32 bits platform since orjson dependency added (eg. RaspberryPI) (Issue #518)


I just tested it myself on a Raspberry Pi. Latest NiceGUI installs fine but is showing the following error on start:

# python3 main.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ImportError: libcblas.so.3: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/test/main.py", line 1, in <module>
    from nicegui import ui
  File "/usr/local/lib/python3.9/dist-packages/nicegui/__init__.py", line 1, in <module>
    from . import elements, globals, ui
  File "/usr/local/lib/python3.9/dist-packages/nicegui/ui.py", line 44, in <module>
    from .elements.scene import Scene as scene
  File "/usr/local/lib/python3.9/dist-packages/nicegui/elements/scene.py", line 9, in <module>
    from .scene_object3d import Object3D
  File "/usr/local/lib/python3.9/dist-packages/nicegui/elements/scene_object3d.py", line 4, in <module>
    import numpy as np
  File "/usr/local/lib/python3.9/dist-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/usr/local/lib/python3.9/dist-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.9 from "/usr/bin/python3"
  * The NumPy version is: "1.24.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: libcblas.so.3: cannot open shared object file: No such file or directory


Which was not the error I've expected. I used the command

sudo apt-get install libatlas-base-dev


to fix it as suggested in https://stackoverflow.com/a/56972113/364388. After that the latest NiceGUI 1.1.11 starts without errors. Please help us reproduce the issue. The system looks like this:

~/test# cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

~/test# uname -mrs
Linux 5.15.84-v7+ armv7l

~/test# getconf LONG_BIT
32

What is different?

**Comment by rbeeli** [comment]
@ghuserlb  Have you tried it? Would `pip` installing `orjson` fail on that 32 bit machine?

**Comment by ghuserlb** [comment]
Hi,

On the Windows 32 setup, the orjson fails with following error :
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error

  Preparing metadata (pyproject.toml) did not run successfully.
  exit code: 1

  [6 lines of output]

  Cargo, the Rust package manager, is not installed or is not on PATH.
  This package requires Rust and Cargo to compile extensions. Install it through
  the system's package manager or via https://rustup.rs/

  Checking for Rust toolchain....
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

Encountered error while generating package metadata.

See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


Then, I looked for this error and found the topic on the unavailability of orjson on 32bits.
I tried also on a 32 bits Linux debian with python 3.7, same error. Since the orjson team seemed to officially not support 32 bits, I did not fight to try to setup the whole Cargo/Rust thing, given these many efforts would likely fail...

**Comment by rbeeli** [comment]
Hm, either we make `orjson` an extra, i.e. `nicegui[orjson]`, or force 32 bit system users to manually install the `nicegui` dependencies after installing `nicegui` itself without dependencies:

`pip install --no-deps nicegui`

I like neither of the solutions. Probably the extra solution is the cleanest, but it's unfortunate that 99% of users would have to do it to profit from `orjson`.

**Comment by rbeeli** [comment]
Actually, there might be a clean solution to hide this from all users:

https://peps.python.org/pep-0508/#environment-markers

https://python-poetry.org/docs/dependency-specification/#using-environment-markers

We should be able to condition `orjson` to its supported platforms, e.g. `platform_machine == x86_64`.

`orjson` [states](https://github.com/ijl/orjson):

It distributes x86_64/amd64, aarch64/armv8, and arm7 wheels for Linux, amd64 and aarch64 wheels for macOS, and amd64 wheels for Windows.

**Comment by ghuserlb** [comment]
Just to mention that I tried the -no-deps pip option but did not get the expected behavior. Orjson still tried to install. My understanding was that -no-deps prevents update of already installed dependencies, but does not prevent from their initial installation.

**Comment by rbeeli** [comment]
> Just to mention that I tried the -no-deps pip option but did not get the expected behavior. Orjson still tried to install. My understanding was that -no-deps prevents update of already installed dependencies, but does not prevent from their initial installation.

Interesting. Well, anyways, the environment markers should provide us with a clean solution so that we only install `orjson` on platforms where it's actually supported.

**Comment by rodja** [comment]
@rbeeli we should try these markers. Looks promising. But first I need a reproduction. I've just installed nicegui and this triggered installing orjson via pip:

```bash
$ python --version
Python 3.9.2
$ python -m pip show orjson
Name: orjson
Version: 3.8.7
Summary: Fast, correct Python JSON library supporting dataclasses, datetimes, and numpy
Home-page: https://github.com/ijl/orjson
Author: ijl <ijl@mailbox.org>
Author-email: ijl <ijl@mailbox.org>
License: Apache-2.0 OR MIT
Location: /usr/local/lib/python3.9/dist-packages
Requires:
Required-by: nicegui
```

@ghuserlb the `--no-deps` worked for me in the past as expected. Installing no dependencies. It need to be written with two dashes `--`. But still, currently NiceGUI won't work without orjson.

**Comment by rodja** [comment]
It looks like there is a prebuild orjson wheel for armv7l:

```
$ sudo python3 -m pip install -v orjson
Using pip 20.3.4 from /usr/lib/python3/dist-packages/pip (python 3.9)
Non-user install because site-packages writeable
Created temporary directory: /tmp/pip-ephem-wheel-cache-ah28xpy3
Created temporary directory: /tmp/pip-req-tracker-9b746pny
Initialized build tracking at /tmp/pip-req-tracker-9b746pny
Created build tracker: /tmp/pip-req-tracker-9b746pny
Entered build tracker: /tmp/pip-req-tracker-9b746pny
Created temporary directory: /tmp/pip-install-9sk_jxy0
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
2 location(s) to search for versions of orjson:
* https://pypi.org/simple/orjson/
* https://www.piwheels.org/simple/orjson/
Fetching project page and analyzing links: https://pypi.org/simple/orjson/
Getting page https://pypi.org/simple/orjson/
Found index url https://pypi.org/simple
Looking up "https://pypi.org/simple/orjson/" in the cache
Request header has "max_age" as 0, cache bypassed
Starting new HTTPS connection (1): pypi.org:443
https://pypi.org:443 "GET /simple/orjson/ HTTP/1.1" 304 0
...
Collecting orjson
  Created temporary directory: /tmp/pip-unpack-b5njb4ev
  Looking up "https://files.pythonhosted.org/packages/59/49/7db5ec40c5289a90f4b15e9c0ea460734332813e75d1e09ef22ae5ffcfbc/orjson-3.8.7-cp39-cp39-manylinux_2_17_armv7l.manylinux2014_armv7l.whl" in the cache
  Current age based on date: 11706
  Ignoring unknown cache-control directive: immutable
  Freshness lifetime from max-age: 365000000
  The response is "fresh", returning cached response
  365000000 > 11706
  Using cached orjson-3.8.7-cp39-cp39-manylinux_2_17_armv7l.manylinux2014_armv7l.whl (149 kB)
  Added orjson from https://files.pythonhosted.org/packages/59/49/7db5ec40c5289a90f4b15e9c0ea460734332813e75d1e09ef22ae5ffcfbc/orjson-3.8.7-cp39-cp39-manylinux_2_17_armv7l.manylinux2014_armv7l.whl#sha256=78604d3acfd7cd502f6381eea0c42281fe2b74755b334074ab3ebc0224100be1 to build tracker '/tmp/pip-req-tracker-9b746pny'
  Removed orjson from https://files.pythonhosted.org/packages/59/49/7db5ec40c5289a90f4b15e9c0ea460734332813e75d1e09ef22ae5ffcfbc/orjson-3.8.7-cp39-cp39-manylinux_2_17_armv7l.manylinux2014_armv7l.whl#sha256=78604d3acfd7cd502f6381eea0c42281fe2b74755b334074ab3ebc0224100be1 from build tracker '/tmp/pip-req-tracker-9b746pny'
Created temporary directory: /tmp/pip-unpack-ni2j0co4
Installing collected packages: orjson

Successfully installed orjson-3.8.7
```

**Comment by rodja** [comment]
So I conclude that Raspberry Pi works fine. Leaves us with 32bit Windows. That will be super hard to integrate in our CI.

@ghuserlb would it be possible for you to checkout the nicegui repo, manually edit the code to not use orjson and then install the checkout with `pip install -e .`? That would make your modified NiceGUI available to your apps. We do the same when developing new features. See https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#setup

**Comment by ghuserlb** [comment]
Ok, I'll see what I can do on my side : try the Rust toolchain just in case, play with a 64bits virtualenv, or fallback to the nicegui repo checkout trick. I'll keep you updated if I find something of interest.
Unless you find a super easy patch for my issue, you can close the issue and stop wasting time on the 1% :)

Thank you all for your help anyway
Regards

**Comment by rbeeli** [comment]
I have created a pull request such that the `json` module falls back to the Python built-in json module if `orjson` is not available, see https://github.com/zauberzeug/nicegui/pull/532.

@rodja Can you test/add the environment markers for `orjson` in `pyproject.toml`?

**Comment by rodja** [comment]
@rbeeli thanks for the effort. I'm not sure it would be good to use a marker for 32 bit architectures. For example on Raspberry Pi it installs fine with 32 bit. And there could be other issues like a missing Rust-compiler #531 which prevent users from using orjson.
I searched long and hard to find a way if we can somehow enable the user to opt-out orjson when installing. But to no avail. It seems like the only possible way would be to make json the default again and define orjson as extra. Such a pitty.

**Comment by rbeeli** [comment]
Or we are ok with 32 bit users using Python `json` by default. If they manage to install `orjson` correctly, `nicegui` would still use it. Since 32 bit users are IMO a very small fraction, that's probably the way I would go. Just make a remark in the installation instructions that people who want to use `orjson` on 32 bit must install it manually, seems not too bad a solution to me.

Also, does `orjson` even officially support 32 bit architectures? Couldn't find any statement. If not, then I suppose it's clear anyways.

**Comment by rodja** [comment]
@rbeeli I guess it would be ok to use default json on 32 bit. But it will come at a maintenance cost. There is the special code you wrote in #532 and custom test infrastructure (because GitHub Actions do not provide 32-bit platforms). Overall it seems like quite a lot of work. Not only in the initial setup, but maintaining it over time. If it is really only affecting a super small minority of users we could simply leave it as it is and point to https://github.com/zauberzeug/nicegui/issues/518#issuecomment-1470623194 for local installation and modification.

**Comment by falkoschindler** [comment]
@rodja Is maintenance and custom test infrastructure that crucial? If we test on 64 bit (covering 99% of the users) and simply cross fingers for 32 bit - wouldn't that be enough? At least it sounds better to me than having 32 bit not covered at all and relying on documentation.
To be clear: I'm in favor of #532. But maybe I'm still missing something.

**Comment by falkoschindler** [comment]
It looks like [#531](https://github.com/zauberzeug/nicegui/issues/531#issuecomment-1471884685) is also related to 32 bit Python, so it would be solved with #532.

**Comment by rbeeli** [comment]
Sounds reasonable to me. There are unit tests for the Python fallback, so at least we know the generated JSON is identical to orjson (for the tested inputs).
